dlang/source/flatbuffers/flatbufferbuilder.d

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+module flatbuffers.flatbufferbuilder;
+
+import flatbuffers.exception;
+import flatbuffers.bytebuffer;
+
+import std.exception;
+import std.traits : isNumeric;
+
+/**
+    Responsible for building up and accessing a FlatBuffer formatted byte
+*/
+final class FlatBufferBuilder
+{
+    /**
+        Create a FlatBufferBuilder with a given initial size.
+        Params:
+            initsize = The initial size to use for the internal buffer.
+    */
+    this(size_t initsize = 32)
+	in{
+		assert(initsize > 0);
+	}body{
+		this(new ByteBuffer(new ubyte[initsize]));
+    }
+
+	this(ByteBuffer buffer){
+		_space = buffer.length;
+		_buffer = buffer;
+	}
+
+    uint offset()
+    {
+        return cast(uint)(_buffer.length - _space);
+    }
+
+    void pad(size_t size)
+    {
+        for (int i = 0; i < size; i++)
+        {
+            --_space;
+            _buffer.put!ubyte(_space, 0x00);
+        }
+    }
+
+    /** Doubles the size of the ByteBuffer, and copies the old data towards
+        the end of the new buffer (since we build the buffer backwards).
+    */
+    void growBuffer()
+    {
+        auto oldBuf = _buffer.data;
+        auto oldBufSize = oldBuf.length;
+        if ((oldBufSize & 0xC0000000) != 0)
+            throw new Exception("FlatBuffers: cannot grow buffer beyond 2 gigabytes.");
+
+		auto newBufSize = oldBufSize >= 32 ? oldBufSize * 2 : 64;
```

Comment:
maybe `auto newBufSize = max(oldBufSize * 2, 64);`  ? Not sure if it is same semantic.